### DISCUSSION TITLE
Make Array's methods chainable

### DISCUSSION THREAD

**Discussion by miv391** [godotengine/godot-proposals#4025] [Scripting]
Change Array's `append`, `append_array`, `duplicate`, `erase`, `insert`, `invert`, `push_back`, `push_front`, `remove`, `resize`, `shuffle`, `sort`, and `sort_custom `methods to return the Array itself. That would allow chaining methods when e.g. a sorted copy of the original array is needed:

```gdscript
var sorted_arr = arr.duplicate().sort()
```

instead of having to do this:
```gdscript
var sorted_arr = arr.duplicate()
sorted.sort()
```

Code would be shorter and in my opinion, also much clearer. In the current way, the first line is a bit misleading, `sorted_arr` is not sorted, it is just a duplicate of the original array. It will be sorted only later.

The downside might be that the programmer then expects that some of those methods return a copy of the original array, but I guess that most programmers would find this proposed way more intuitive. This is also how e.g. Javascript works, so I guess this is how most programmers with experience with other languages would expect these methods to work.

**Comment by miv391** [comment]
Actually my example ended a bit unfortunate, as `duplicate` is pretty much only method of those that actually returns an array. But it still is not possible to chain it with other methods.

**Comment by Mickeon** [comment]
I can at *most* only speculate that the reason this is not the case is to make these methods much clearer to beginner users. A long text of methods that do not imply what is actually being returned can be somewhat tricky. It may also be some sort of performance benefit(?).

Chaining can be easily done with **Strings**, but those are immutable, and a whole different String is returned when operating on them.

To be honest, I find the inability to chain these methods to be most apparent during initialization. I often want to resize Arrays after defining it to avoid appending later, or storing a slight variation of the same array, and that may require just a few extra more lines that it feels necessary.

**Comment by Calinou** [comment]
Using in-place methods that modify the array can be preferable for performance reasons, especially with larger arrays.

**Reply by theraot** [reply]
Given that OP says "return the Array *itself*" I believe the idea would be to do those operations in place, but also have them return a reference to the same array they were called on. That way we can chain them, and if we want to duplicate the array, we can chain with `duplicate()` too.

**Reply by miv391** [reply]
Yes, sorting in place does not prevent returning reference to array itself and chaining methods. For example in Typescript you can get top 3 values like this:

```typescript
// Typescript
const arr = [6,8,3,5,2,4,1]
const top3 = arr.sort().reverse().slice(0, 3)
```
Typescript sorts and reverses array in place just like GDScript. So at the end `arr` is sorted and reversed: `[8, 6, 5, 4, 3, 2, 1]` 

The same operation in GDScript is more cumbersome. The end result is exactly the same. `arr` is sorted and reversed and `top3` contains three largest values.

```gdscript
# GDScript now
var arr = [6,8,3,5,2,4,1]
arr.sort()
arr.reverse()
var top3 = arr.slice(0, 3)

# GDScript if chaining were possible
var arr = [6,8,3,5,2,4,1]
var top3 = arr.sort().reverse().slice(0, 3)
```

The point of this proposal is to make array operations shorter and more fluent.

**Reply by Mickeon** [reply]
Perhaps these could be their own new methods that ONLY return a new Array, but named as `sorted()` , `reversed()`, `sliced()`... to make sense reading them.

However, I can easily see they would bloat the Array documentation and potentially be very, very confusing.

**Reply by miv391** [reply]
As Typescript (with huge user base) makes sense, I'm sure GDScript working at the _exactly_ same way would make sense too. I don't think methods returning copies would be very useful, `duplicate` is enough. They would just cause unintended performance problems when copies would be made even when not needed.

**Comment by theraot** [comment]
> The downside might be that the programmer then expects that some of those methods return a copy of the original array, but I guess that most programmers would find this proposed way more intuitive. This is also how e.g. Javascript works, so I guess this is how most programmers with experience with other languages would expect these methods to work.

Intuition is not build on nothing. Yes, some of it is knowledge transfer from other programming languagess. However, you should also consider natural languages. If I see `array.sort()` I expect that this method modifies the array in place. And if I see `array.sorted()` I expect that this method returns a sorted copy. I'm aware it is not a clear cut like that for every case, however, I'm trying to point that good naming is important for that "intuition" to work.
